File: src/golang/p200to299/211.添加与搜索单词-数据结构设计.go
package p200to299

/*
 * @lc app=leetcode.cn id=211 lang=golang
 *
 * [211] 添加与搜索单词 - 数据结构设计
 *
 * https://leetcode-cn.com/problems/add-and-search-word-data-structure-design/description/
 *
 * algorithms
 * Medium (45.15%)
 * Likes:    137
 * Dislikes: 0
 * Total Accepted:    13K
 * Total Submissions: 28.6K
 * Testcase Example:  '["WordDictionary","addWord","addWord","addWord","search","search","search","search"]\n' +
  '[[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]'
 *
 * 设计一个支持以下两种操作的数据结构：
 *
 * void addWord(word)
 * bool search(word)
 *
 *
 * search(word) 可以搜索文字或正则表达式字符串，字符串只包含字母 . 或 a-z 。 . 可以表示任何一个字母。
 *
 * 示例:
 *
 * addWord("bad")
 * addWord("dad")
 * addWord("mad")
 * search("pad") -> false
 * search("bad") -> true
 * search(".ad") -> true
 * search("b..") -> true
 *
 *
 * 说明:
 *
 * 你可以假设所有单词都是由小写字母 a-z 组成的。
 *
*/

/**
 * @File    :   211.添加与搜索单词-数据结构设计.go
 * @Time    :   2020/07/18 19:14:19
 * @Author  :   wylu
 * @Version :   1.0
 * @Contact :   15wylu@gmail.com
 * @License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
 * @Desc    :
 */
// @lc code=start

// Trie211 - Prifix Tree
type Trie211 struct {
	links [26]*Trie211
	isEnd bool
}

// WordDictionary - word dictionary
type WordDictionary struct {
	trie *Trie211
}

// Constructor211 - Initialize your data structure here.
func Constructor211() WordDictionary {
	return WordDictionary{
		trie: &Trie211{},
	}
}

// AddWord - Adds a word into the data structure.
func (wd *WordDictionary) AddWord(word string) {
	tr := wd.trie
	for i := 0; i < len(word); i++ {
		ch := word[i] - 'a'
		if tr.links[ch] == nil {
			tr.links[ch] = &Trie211{}
		}
		tr = tr.links[ch]
	}
	tr.isEnd = true
}

// Search - Returns if the word is in the data structure.
// A word could contain the dot character '.' to represent any one letter.
func (wd *WordDictionary) Search(word string) bool {
	return wd.trie.Search211(word, 0)
}

// Search211 - search a word
func (tr *Trie211) Search211(word string, k int) bool {
	for i := k; i < len(word); i++ {
		if word[i] == '.' {
			for j := 0; j < len(tr.links); j++ {
				if tr.links[j] != nil && tr.links[j].Search211(word, i+1) {
					return true
				}
			}
			return false
		}

		ch := word[i] - 'a'
		if tr.links[ch] == nil {
			return false
		}
		tr = tr.links[ch]
	}
	return tr.isEnd
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddWord(word);
 * param_2 := obj.Search(word);
 */
// @lc code=end

// // WordDictionary - word dictionary
// type WordDictionary struct {
// 	trie *Trie211
// }

// // Constructor211 - Initialize your data structure here.
// func Constructor211() WordDictionary {
// 	return WordDictionary{
// 		trie: &Trie211{},
// 	}
// }

// // AddWord - Adds a word into the data structure.
// func (wd *WordDictionary) AddWord(word string) {
// 	wd.trie.Insert211(word)
// }

// // Search - Returns if the word is in the data structure.
// // A word could contain the dot character '.' to represent any one letter.
// func (wd *WordDictionary) Search(word string) bool {
// 	return wd.trie.Search211(word)
// }

// // Trie211 - Prifix Tree
// type Trie211 struct {
// 	links [26]*Trie211
// 	isEnd bool
// }

// // Insert211 - insert a word
// func (tr *Trie211) Insert211(word string) {
// 	for i := 0; i < len(word); i++ {
// 		ch := word[i] - 'a'
// 		if tr.links[ch] == nil {
// 			tr.links[ch] = &Trie211{}
// 		}
// 		tr = tr.links[ch]
// 	}
// 	tr.isEnd = true
// }

// // Search211 - search a word
// func (tr *Trie211) Search211(word string) bool {
// 	for i := 0; i < len(word); i++ {
// 		if word[i] == '.' {
// 			flag := false
// 			for j := 0; j < len(tr.links) && !flag; j++ {
// 				if tr.links[j] == nil {
// 					continue
// 				}
// 				flag = tr.links[j].Search211(string(word[i+1:]))
// 			}
// 			return flag
// 		}

// 		ch := word[i] - 'a'
// 		if tr.links[ch] == nil {
// 			return false
// 		}
// 		tr = tr.links[ch]
// 	}
// 	return tr.isEnd
// }

File: src/python/p200to299/211.添加与搜索单词-数据结构设计.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@File    :   211.添加与搜索单词-数据结构设计.py
@Time    :   2021/10/19 14:02:49
@Author  :   wylu
@Version :   1.0
@Contact :   15wylu@gmail.com
@License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
@Desc    :
"""

#
# @lc app=leetcode.cn id=211 lang=python3
#
# [211] 添加与搜索单词 - 数据结构设计
#
# https://leetcode-cn.com/problems/design-add-and-search-words-data-structure/description/
#
# algorithms
# Medium (49.79%)
# Likes:    328
# Dislikes: 0
# Total Accepted:    37.8K
# Total Submissions: 75.9K
# Testcase Example:
# '["WordDictionary","addWord","addWord","addWord","search","search","search","search"]\n'
# + '[[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]'
#
# 请你设计一个数据结构，支持 添加新单词 和 查找字符串是否与任何先前添加的字符串匹配 。
#
# 实现词典类 WordDictionary ：
#
#
# WordDictionary() 初始化词典对象
# void addWord(word) 将 word 添加到数据结构中，之后可以对它进行匹配
# bool search(word) 如果数据结构中存在字符串与 word 匹配，则返回 true ；否则，返回  false 。word 中可能包含一些
# '.' ，每个 . 都可以表示任何一个字母。
#
#
#
#
# 示例：
#
#
# 输入：
#
# ["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
# [[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
# 输出：
# [null,null,null,null,false,true,true,true]
#
# 解释：
# WordDictionary wordDictionary = new WordDictionary();
# wordDictionary.addWord("bad");
# wordDictionary.addWord("dad");
# wordDictionary.addWord("mad");
# wordDictionary.search("pad"); // return False
# wordDictionary.search("bad"); // return True
# wordDictionary.search(".ad"); // return True
# wordDictionary.search("b.."); // return True
#
#
#
#
# 提示：
#
#
# 1 <= word.length <= 500
# addWord 中的 word 由小写英文字母组成
# search 中的 word 由 '.' 或小写英文字母组成
# 最多调用 50000 次 addWord 和 search
#
#
#


# @lc code=start
class Trie:
    def __init__(self) -> None:
        self.children = {}
        self.flag = False

    def insert(self, word: str) -> None:
        cur = self
        for ch in word:
            if ch not in cur.children:
                cur.children[ch] = Trie()
            cur = cur.children[ch]
        cur.flag = True

    def search(self, word: str, s: int) -> bool:
        cur = self
        for i in range(s, len(word)):
            ch = word[i]
            if ch == '.':
                for _, child in cur.children.items():
                    if child.search(word, i + 1):
                        return True
                return False

            if ch not in cur.children:
                return False
            cur = cur.children[ch]
        return cur.flag


class WordDictionary:
    def __init__(self):
        self.trie = Trie()

    def addWord(self, word: str) -> None:
        self.trie.insert(word)

    def search(self, word: str) -> bool:
        return self.trie.search(word, 0)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
# @lc code=end
